File: ts/src/embed/similarity.ts
export function dotProduct(a: number[], b: number[]) {
    return a.map((value, index) => value * b[index]).reduce((a, b) => a + b, 0);
}

export function cosineSimilarity(a: number[], b: number[]) {
    const product = dotProduct(a, b);
    const aMagnitude = Math.sqrt(a.map(value => value * value).reduce((a, b) => a + b, 0));
    const bMagnitude = Math.sqrt(b.map(value => value * value).reduce((a, b) => a + b, 0));
    return product / (aMagnitude * bMagnitude);
}

const list1 = [1, 2, 3];
const list2 = [1, 2, 3];
const list3 = [-1, -2, -3];
// console.log(dotProduct(list1, list2));
// console.log(cosineSimilarity(list1, list3));

File: py/src/embed/similarity.py
def dotProduct(embedding1: list, embedding2: list):
    return sum([embedding1[i] * embedding2[i] for i in range(len(embedding1))])

def cosineSimilarity(embedding1: list, embedding2: list):
    dotProductValue = dotProduct(embedding1, embedding2)
    magnitude1 = dotProduct(embedding1, embedding1) ** 0.5
    magnitude2 = dotProduct(embedding2, embedding2) ** 0.5
    return dotProductValue / (magnitude1 * magnitude2)

list1 = [1, 2, 3]
list2 = [1, 2, 3]
list3 = [-1, -2, -3]

# print(cosineSimilarity(list1, list2))
# print(cosineSimilarity(list1, list3))